syntax = "proto3";

package brain.envelope.v1;

import "google/protobuf/timestamp.proto";
import "google/protobuf/any.proto";

// Envelope kind for cross-service and cross-layer communication.
enum EnvelopeKind {
  ENVELOPE_KIND_UNSPECIFIED = 0;
  ENVELOPE_KIND_COMMAND     = 1;
  ENVELOPE_KIND_EVENT       = 2;
  ENVELOPE_KIND_RESULT      = 3;
  ENVELOPE_KIND_STREAM      = 4; // for future use
}

// Error category for typed failure handling.
enum ErrorCategory {
  ERROR_CATEGORY_UNSPECIFIED = 0;
  ERROR_CATEGORY_VALIDATION  = 1;
  ERROR_CATEGORY_CONFLICT    = 2;
  ERROR_CATEGORY_NOT_FOUND   = 3;
  ERROR_CATEGORY_POLICY      = 4;
  ERROR_CATEGORY_DEPENDENCY  = 5;
  ERROR_CATEGORY_INTERNAL    = 6;
}

// Standard envelope metadata propagated end-to-end.
message EnvelopeMeta {
  string envelope_id                  = 1;
  string trace_id                     = 2;
  string parent_id                    = 3;
  EnvelopeKind kind                   = 4;
  google.protobuf.Timestamp timestamp = 5;
  string source                       = 6;
  string principal                    = 7;
}                                     

// Structured error details returned by services.
message ErrorDetail {
  string code = 1;
  string message = 2;
  ErrorCategory category = 3;
  bool retryable = 4;
  map<string, string> metadata = 5;
}

// Command envelope used for intentful invocations.
//
// Payload is implementation-defined and encoded as google.protobuf.Any.
message CommandEnvelope {
  EnvelopeMeta metadata = 1;
  google.protobuf.Any payload = 2;
  repeated ErrorDetail errors = 3;
}

// Event envelope used for emitted system or external events.
//
// Payload is implementation-defined and encoded as google.protobuf.Any.
message EventEnvelope {
  EnvelopeMeta metadata = 1;
  google.protobuf.Any payload = 2;
  repeated ErrorDetail errors = 3;
}

// Result envelope used for responses to command/event handling.
//
// Payload is implementation-defined and encoded as google.protobuf.Any.
message ResultEnvelope {
  EnvelopeMeta metadata = 1;
  google.protobuf.Any payload = 2;
  repeated ErrorDetail errors = 3;
}
